### PR TITLE
Support openbsd

### PR DESCRIPTION
**Pull request by flub**
This makes Quinn compile on openbsd and the test suite passes as well.

As far as I understand ECN is not supported as part of a control message on recvmsg.  So I disabled the ECN tests for it.

### CODE REVIEW

File: quinn-udp/src/unix.rs [hunk 1]
```diff
@@ -138,12 +145,19 @@ impl UdpSocketState {
...
+                    &*io,
+                    libc::IPPROTO_IPV6,
+                    libc::IPV6_DONTFRAG,
+                    OPTION_ON,
+                )?;
+            }
+            #[cfg(target_os = "openbsd")]
+            {
+                may_fragment |= true;
+            }
```
Comment (Ralith) [inline quinn-udp/src/unix.rs:160]: Why are we skipping this? OpenBSD seems to define the symbol:

https://github.com/openbsd/src/blob/35659bf2453d7eeb12fd24089cf668ca13361c18/sys/netinet6/in6.h#L336

File: quinn-udp/src/unix.rs [hunk 2]
```diff
@@ -293,7 +307,7 @@ fn send(
     }
 }
 
-#[cfg(any(target_os = "macos", target_os = "ios"))]
+#[cfg(any(target_os = "macos", target_os = "ios", target_os = "openbsd"))]
```
Comment (Ralith) [inline quinn-udp/src/unix.rs:310]: `IP_RECVDSTADDR` seems to be defined in OpenBSD; should we add it to the cases where the `encode_ip` argument to `prepare_msg` is true below?

File: quinn-udp/src/unix.rs (hunk 1 above)
Comment (flub) [reply quinn-udp/src/unix.rs:160]: The libc crate does not have this symbol, I would prefer to have something sub-optimal working that can later be improved than wait until this is fully supported.

**Review by djc — APPROVED** [review]
Seems reasonable to merge this even without ECN support for now.

File: quinn-udp/src/unix.rs (hunk 2 above)
Comment (flub) [reply quinn-udp/src/unix.rs:310]: good catch, yes.  testing that shows it also works.

**Comment by flub** [comment]
> OpenBSD does seem to define `IPV6_RECVTCLASS`: https://github.com/openbsd/src/blob/35659bf2453d7eeb12fd24089cf668ca13361c18/sys/netinet6/in6.h#L330C9-L330C24

Indeed, and the socket option is being set as well still.  Let me try again if I can get ENC to work for just IPv6.

**Comment by flub** [comment]
> Thanks for working on this!
> 
> > As far as I understand ECN is not supported as part of a control message on recvmsg. So I disabled the ECN tests for it.
> 
> OpenBSD does seem to define `IPV6_RECVTCLASS`: https://github.com/openbsd/src/blob/35659bf2453d7eeb12fd24089cf668ca13361c18/sys/netinet6/in6.h#L330C9-L330C24

Ok, ECN now works and is tested on IPv6 but not IPv4.  I've had to split some tests for that.  I left the existing ecn_ipv6 test unmodified in the end but renamed it rather than fully splitting it in two.  It still makes sense on it's own.

**Comment by flub** [comment]
This will close #1469 I believe.

File: quinn-udp/src/unix.rs (hunk 1 above)
Comment (Ralith) [reply quinn-udp/src/unix.rs:165]: We don't need to wait for `libc` to define the constant for us; it's just a number. Let's define the constant ourselves, ideally with a TODO pointing to a libc issue or PR.

File: quinn-udp/src/unix.rs [hunk 3]
```diff
@@ -579,7 +603,11 @@ fn decode_recv(
     for cmsg in cmsg_iter {
         match (cmsg.cmsg_level, cmsg.cmsg_type) {
             // FreeBSD uses IP_RECVTOS here, and we can be liberal because cmsgs are opt-in.
-            (libc::IPPROTO_IP, libc::IP_TOS) | (libc::IPPROTO_IP, libc::IP_RECVTOS) => unsafe {
+            (libc::IPPROTO_IP, libc::IP_TOS) => unsafe {
+                ecn_bits = cmsg::decode::<u8, libc::cmsghdr>(cmsg);
+            },
+            #[cfg(not(target_os = "openbsd"))]
+            (libc::IPPROTO_IP, libc::IP_RECVTOS) => unsafe {
```
Comment (Ralith) [inline quinn-udp/src/unix.rs:610]: The comment mentioning `IP_RECVTOS` should probably remain attached to the case that matches it

File: quinn-udp/tests/tests.rs [hunk 4]
```diff
@@ -31,6 +33,46 @@ fn basic() {
...
+                segment_size: None,
+                src_ip: None,
+            },
+        );
+    }
+}
+
+#[test]
+#[cfg(not(target_os = "openbsd"))]
+fn enc_v6_dualstack() {
```
Comment (Ralith) [inline quinn-udp/tests/tests.rs:75]: typo
```suggestion
fn ecn_v6_dualstack() {
```

File: quinn/src/tests.rs [hunk 5]
```diff
@@ -532,6 +532,7 @@ fn run_echo(args: EchoArgs) {
             // platforms in the doc comments of the various `local_ip` functions.
```
Comment (Ralith) [inline quinn/src/tests.rs:532]: Please update docs as discussed in this above comment.

**Review by Ralith — CHANGES_REQUESTED** [review]

File: quinn/src/tests.rs (hunk 5 above)
Comment (Ralith) [reply quinn/src/tests.rs:532]: Note these docs are being centralized in https://github.com/quinn-rs/quinn/pull/1868.

File: quinn-udp/src/unix.rs (hunk 1 above)
Comment (flub) [reply quinn-udp/src/unix.rs]: Cool, I've updated this now to just define the constant.  All still works.

**Comment by flub** [comment]
It is not clear to me if the failing test is flaky or due to something I did, any hints?

**Review by Ralith — APPROVED** [review]
LGTM, thanks!

**Comment by flub** [comment]
@Ralith thanks for the prodding, definitely better than the first version!
